exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java

```diff
@@ -176,81 +177,103 @@ protected void doOnMatch(RelOptRuleCall call, DrillFilterRel filterRel, DrillPro
-            optimizerContext.getFunctionRegistry(),
-            errors);
+    List<String> newFiles = Lists.newArrayList();
+    long numTotal = 0; // total number of partitions
+    int batchIndex = 0;
+    String firstLocation = null;
+
+    // Outer loop: iterate over a list of batches of PartitionLocations
+    for (List<PartitionLocation> partitions : descriptor) {
+      numTotal += partitions.size();
+      logger.debug("Evaluating partition pruning for batch {}", batchIndex);
+      if (batchIndex == 0) { // save the first location in case everything is pruned
+        firstLocation = partitions.get(0).getEntirePartitionLocation();
       }
+      final NullableBitVector output = new NullableBitVector(MaterializedField.create("", Types.optional(MinorType.BIT)), allocator);
+      final VectorContainer container = new VectorContainer();
+
+      try {
+        final ValueVector[] vectors = new ValueVector[descriptor.getMaxHierarchyLevel()];
+          for (int partitionColumnIndex : BitSets.toIter(partitionColumnBitSet)) {
+          SchemaPath column = SchemaPath.getSimplePath(fieldNameMap.get(partitionColumnIndex));
+          MajorType type = descriptor.getVectorType(column, settings);
+          MaterializedField field = MaterializedField.create(column, type);
+          ValueVector v = TypeHelper.getNewVector(field, allocator);
+          v.allocateNew();
+          vectors[partitionColumnIndex] = v;
+          container.add(v);
+        }
 
-      if (errors.getErrorCount() != 0) {
-        logger.warn("Failure while materializing expression [{}].  Errors: {}", expr, errors);
-      }
+        // populate partition vectors.
+        descriptor.populatePartitionVectors(vectors, partitions, partitionColumnBitSet, fieldNameMap);
+
+        // materialize the expression
+        logger.debug("Attempting to prune {}", pruneCondition);
+        final LogicalExpression expr = DrillOptiq.toDrill(new DrillParseContext(settings), scanRel, pruneCondition);
+        final ErrorCollectorImpl errors = new ErrorCollectorImpl();
+
+        LogicalExpression materializedExpr = ExpressionTreeMaterializer.materialize(expr, container, errors, optimizerContext.getFunctionRegistry());
+        // Make sure pruneCondition's materialized expression is always of BitType, so that
+        // it's same as the type of output vector.
+        if (materializedExpr.getMajorType().getMode() == TypeProtos.DataMode.REQUIRED) {
+          materializedExpr = ExpressionTreeMaterializer.convertToNullableType(
+              materializedExpr,
+              materializedExpr.getMajorType().getMinorType(),
+              optimizerContext.getFunctionRegistry(),
+              errors);
+        }
 
-      output.allocateNew(partitions.size());
-      InterpreterEvaluator.evaluate(partitions.size(), optimizerContext, container, output, materializedExpr);
-      int record = 0;
+        if (errors.getErrorCount() != 0) {
```

Comment:
Actually, the ExpressionTreeMaterializer.visitSchemaPath() needs access to the VectorContainer, so doing this materialization outside the loop is not an option since the VectorContainer is initialized inside the loop.  But we can do this only once by putting it inside the if (batchIndex == 0)  condition.  In fact, the container itself should be created only once and cleared in each iteration.  I will make that change.